privateSBOMExchange/src/petra/lib/models/__init__.py

```diff
@@ -646,3 +696,147 @@ def build_sbom_tree(parser:SBOMParser):
     return root
 
 
+class GetTargetNodes:
+    """Visitor that collects the data and hash of each node.
+    This is used to test for membership"""
+
+    def __init__(self):
+        # Initialize a list to store the hashes of visited nodes
+        self.hashes = []
+
+    def visit_field_node(self, node):
+        # Append the hash if it exists, otherwise raise error
+        try:
+            node_hash = node.hash.hex()
+            self.hashes.append(node_hash)
+        except AttributeError:
+            pass
+    def visit_complex_node(self, node):
+        # Add the complex node's hash, if available
+        try:
+            node_hash = node.hash.hex()
+            self.hashes.append(node_hash)
+        except AttributeError:
+            pass
+
+        # Visit each child node
+        for child in node.children:
+            child.accept(self)
+
+    def visit_sbom_node(self, node):
+        # Add the SBOM node's hash, if available, and visit children
+        try:
+            node_hash = node.hash.hex()
+            self.hashes.append(node_hash)
+        except AttributeError:
+            pass
+        # Visit each child node
+        for child in node.children:
+            child.accept(self)
+
+    def get_hashes(self):
+        """
+        Returns the collected hashes to the caller.
+        Called after the tree traversal is complete.
+        """
+        return self.hashes
+
+
+def get_membership_proof(root, target_hash):
+    """
+    Generate a membership proof for a target hash within an SBOM tree.
+
+    Parameters:
+    root: The SBOM tree instance.
+    target_hash (str or bytes): The hash of the target item to locate in the tree.
+
+    Returns:
+    list: A list representing the path taken to locate the target hash in the tree.
+          If the root node itself matches the target hash, returns an empty path.
+
+    Raises:
+    Exception: If the root is not an instance of the expected SbomNode class.
+    """
+    target_hash = target_hash.hex() if isinstance(target_hash, bytes) else target_hash
+
+    def traverse(node, path):
+        """
+        Recursively traverse the tree from a given node to find the target hash.
+
+        Parameters:
+        node (SbomNode): The current node being traversed.
+        path (list): The path recorded up to the current node.
+
+        Returns:
+        list or None: If the target hash is found, returns the path to it;
+                      otherwise, returns None.
+        """
+        for child in node.children:
+            if child.hash.hex() == target_hash:
```

Comment:
Why not compare raw bytes directly, instead of spending extra cycles to encode hashes into hex strings?